source/SkiaSharp.Views.Forms/SkiaSharp.Views.Forms.Native.Shared/SKCanvasViewRendererBase.cs

```diff
@@ -0,0 +1,129 @@
+using SKNativePaintSurfaceEventArgs = SkiaSharp.Views.UWP.SKPaintSurfaceEventArgs;
+#endif
+
+namespace SkiaSharp.Views.Forms
+{
+	public abstract class SKCanvasViewRendererBase<TFormsView, TNativeView> : ViewRenderer<TFormsView, TNativeView>
+		where TFormsView : SKFormsView
+		where TNativeView : SKNativeView
+	{
+		protected override void OnElementChanged(ElementChangedEventArgs<TFormsView> e)
+		{
+			if (e.OldElement != null)
+			{
+				var oldController = (ISKCanvasViewController)e.OldElement;
+
+				// unsubscribe from events
+				oldController.SurfaceInvalidated -= OnSurfaceInvalidated;
+				oldController.GetCanvasSize -= OnGetCanvasSize;
+			}
+
+			if (Control != null)
+			{
+				var control = Control;
+				control.PaintSurface -= OnPaintSurface;
+			}
+
+			if (e.NewElement != null)
+			{
+				var newController = (ISKCanvasViewController)e.NewElement;
+
+				// create the native view
+				var view = CreateNativeControl();
+				view.IgnorePixelScaling = e.NewElement.IgnorePixelScaling;
+				view.PaintSurface += OnPaintSurface;
+				SetNativeControl(view);
+
+				// subscribe to events from the user
+				newController.SurfaceInvalidated += OnSurfaceInvalidated;
+				newController.GetCanvasSize += OnGetCanvasSize;
+
+				// paint for the first time
+				OnSurfaceInvalidated(newController, EventArgs.Empty);
+			}
+
+			base.OnElementChanged(e);
+		}
+		
+#if __ANDROID__
+		protected override TNativeView CreateNativeControl()
+		{
+			return (TNativeView)Activator.CreateInstance(typeof(TNativeView), new[] { Context });
+		}
+#else
+		protected virtual TNativeView CreateNativeControl()
+		{
+			return (TNativeView)Activator.CreateInstance(typeof(TNativeView));
+		}
+#endif
+
+		protected override void OnElementPropertyChanged(object sender, PropertyChangedEventArgs e)
+		{
+			base.OnElementPropertyChanged(sender, e);
+
+			if (e.PropertyName == nameof(SKFormsView.IgnorePixelScaling))
+			{
+				Control.IgnorePixelScaling = Element.IgnorePixelScaling;
+			}
+		}
+
+		protected override void Dispose(bool disposing)
+		{
+			// detach all events before disposing
+			var controller = (ISKCanvasViewController)Element;
+			if (controller != null)
+			{
+				controller.SurfaceInvalidated -= OnSurfaceInvalidated;
+				controller.GetCanvasSize -= OnGetCanvasSize;
+			}
+
+			var control = Control;
+			if (control != null)
+			{
+				control.PaintSurface -= OnPaintSurface;
+			}
+
+			base.Dispose(disposing);
+		}
+
+		private void OnPaintSurface(object sender, SKNativePaintSurfaceEventArgs e)
+		{
+			var controller = Element as ISKCanvasViewController;
+
+			// the control is being repainted, let the user know
+			controller?.OnPaintSurface(new SKPaintSurfaceEventArgs(e.Surface, e.Info));
+		}
+
+		private void OnSurfaceInvalidated(object sender, EventArgs eventArgs)
+		{
+			// repaint the native control
+#if __IOS__
+			Control.SetNeedsDisplay();
+#else
+			Control.Invalidate();
+#endif
+		}
```

Comment:
In our current beta, we had several crashes, because "OnSurfaceInvalidated" was called by a background thread.
Maybe it is better to post these on the main thread?